ghmirror/app/__init__.py

```diff
@@ -0,0 +1,103 @@
+import logging
+import requests
+
+import flask
+
+from ghmirror.app.response import MirrorResponse
+from ghmirror.data_structures.cache import RequestsCache
+from ghmirror.data_structures.cache import StatsCache
+
+
+logging.basicConfig(level=logging.INFO,
+                    format='%(asctime)-15s %(message)s')
+
+LOG = logging.getLogger(__name__)
+GH_API = 'https://api.github.com'
+
+APP = flask.Flask(__name__)
+
+
+@APP.route('/healthz', methods=["GET"])
+def healthz():
+    return flask.Response('OK')
+
+
+@APP.route('/stats', methods=["GET"])
+def stats():
+    stats_cache = StatsCache()
+
+    return flask.jsonify(
+        {
+            'cache_hit': stats_cache.hits,
+            'cache_miss': stats_cache.misses,
+        }
+    )
+
+
+@APP.route('/', defaults={'path': ''})
+@APP.route('/<path:path>', methods=["GET", "POST", "PUT", "PATCH", "DELETE"])
+def ghmirror(path):
+    url = f'{GH_API}/{path}'
+
+    if flask.request.args:
+        url += '?'
+        for key, value in flask.request.args.items():
+            url += f'{key}={value}&'
+        url = url.rstrip('&')
+
+    headers = {}
+    authorization = flask.request.headers.get('Authorization')
+    if authorization is not None:
+        headers['Authorization'] = authorization
+
+    if flask.request.method != 'GET':
+        LOG.info('[%s] BYPASS %s', flask.request.method, url)
+
+        resp = requests.request(flask.request.method,
+                                url=url,
+                                headers=headers,
+                                data=flask.request.data)
+
+        return flask.Response(resp.content,
+                              resp.status_code)
+
+    stats_cache = StatsCache()
+    requests_cache = RequestsCache()
+
+    if url in requests_cache:
+        last_modified = requests_cache[url].headers.get('Last-Modified')
+        etag = requests_cache[url].headers.get('ETag')
+        if last_modified is not None:
+            headers['If-Modified-Since'] = last_modified
+        elif etag is not None:
+            headers['If-None-Match'] = etag
+
+    resp = requests.request(flask.request.method,
+                            url=url,
+                            headers=headers)
+
```

Comment:
@jmelis  mentioned "We should handle 500's (apparently GitHub returns 500 to the conditional-request sometimes)" although I don't know how can you differentiate a 500 that is due to the conditional request mechanism from a 500 from the whole github api

Reply:
yes, but I have never seen that happening though...